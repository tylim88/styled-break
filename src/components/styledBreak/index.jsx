import styled, { css } from 'styled-components'

const mediaParser = (
	direction = '',
	sortedBreakpoints = {},
	targetPoint = ''
) => {
	const targetPointValue = sortedBreakpoints[targetPoint]

	const getMaxWidth = targetPointValue => {
		const values = Object.values(sortedBreakpoints)
		const result = values[values.indexOf(targetPointValue) + 1]
		return result ? result - 0.02 : 999999
	}

	const minMedia = `@media (min-width: ${targetPointValue}px)`
	const maxMedia = `@media (max-width: ${getMaxWidth(targetPointValue)}px)`
	switch (direction) {
		case '':
			return minMedia
		case 'm':
			return maxMedia
		case 'o':
			return `${minMedia} and ${maxMedia}`
		default: {
			const targetPointValue2 = sortedBreakpoints[direction]
			if (targetPointValue2) {
				return `${minMedia} and (max-width: ${getMaxWidth(
					targetPointValue2
				)}px)`
			} else {
				return minMedia
			}
		}
	}
}

const responsiveStyledGenerator = config => {
	const { breakpoints, sLevel } = config

	const sLevel_ = sLevel || 1

	const sortedBreakpoints = Object.keys(breakpoints)
		.sort((a, b) => {
			return breakpoints[a] - breakpoints[b]
		})
		.reduce((acc, key) => {
			acc[key] = breakpoints[key]
			return acc
		}, {})

	const cssR = (styledCss = '') => {
		const type = typeof styledCss
		if (type === 'string' || Array.isArray(styledCss)) {
			return styledCss
		} else if (type === 'object' && styledCss) {
			let cssString = []
			for (const prop in styledCss) {
				const split = prop.split('_')
				const targetPoint = split[0]
				if (
					styledCss[prop] !== undefined &&
					sortedBreakpoints[targetPoint] !== undefined
				) {
					const direction = split[1] || ''
					cssString = [
						...cssString,
						`
						${mediaParser(direction, sortedBreakpoints, targetPoint)} {`,
						styledCss[prop],
						`
					}
					`,
					]
				}
			}
			return cssString
		} else {
			return ''
		}
	}

	const isComponentHtml = component =>
		typeof component === 'string' ? styled[component] : styled(component)

	const specificityWrapper = (styledCss = '', level = sLevel_) => css`
		${'&'.repeat(level)} {
			${cssR(styledCss)}
		}
	`

	const styledR = comp => (styledCss = '', level = sLevel_) => {
		return isComponentHtml(comp)`
	 ${specificityWrapper(styledCss, level)}
	 `
	}
	const styledHOC = comp => (level = sLevel_) => {
		return isComponentHtml(comp)`
			${props => {
				const { styledCss } = props
				return specificityWrapper(styledCss, level)
			}}
		`
	}

	return { cssR, styledR, styledHOC }
}

export default responsiveStyledGenerator
