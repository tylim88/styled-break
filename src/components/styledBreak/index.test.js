import styledBreak, { getMaxWidth, getMediaQuery } from './index'

const gap = 0.02
const infinity = 999999

const checkLength = (arr1, arr2) => {
	it('check array length', () => {
		expect.assertions(1)
		expect(arr1.length).toBe(arr2.length)
	})
}

describe('test utilities', () => {
	describe('test getMaxWidth', () => {
		const minWidthArr = [0, 576, 768, 992, 1200]
		const minWidthArr2 = [...minWidthArr]
		minWidthArr2.shift()
		const maxWidthArr = minWidthArr2.map(num => num - gap)
		maxWidthArr.push(infinity)

		checkLength(minWidthArr, maxWidthArr)

		minWidthArr.forEach((minWidth, i) => {
			it(`test ${minWidth}`, () => {
				expect.assertions(1)
				const maxWidth = getMaxWidth(minWidthArr, minWidth)
				expect(maxWidth).toBe(maxWidthArr[i])
			})
		})
	})

	describe('test getMediaQuery', () => {
		const sortedBreakpoint = {
			xs: 0,
			sm: 576,
			md: 768,
			lg: 992,
			xl: 1200,
		}

		const modeArr = [
			'xs',
			'xs_n',
			'sm_m',
			'sm_o',
			'md_lg',
			'md_md',
			'lg_sm',
			'lg_jshdjhskajd',
			'xl_n',
			'xl_m',
		]

		const minOnly = width => `@media (min-width: ${width}px)`
		const maxOnly = width => `@media (max-width: ${width}px)`
		const minMax = (width1, width2) =>
			`@media (min-width: ${width1}px) and (max-width: ${width2}px)`

		const answer = [
			minOnly(0),
			minOnly(0),
			maxOnly(768 - gap),
			minMax(576, 768 - 0.02),
			minMax(768, 1200 - gap),
			minMax(768, 992 - gap),
			minMax(992, 768 - gap),
			minOnly(992),
			minOnly(1200),
			maxOnly(infinity),
		]

		checkLength(modeArr, answer)

		modeArr.forEach((mode, i) => {
			it(`test ${mode}`, () => {
				expect.assertions(1)
				const [targetPoint, direction] = mode.split('_')
				const targetPoint_ = targetPoint || ''
				const mediaQuery = getMediaQuery(
					direction,
					sortedBreakpoint,
					targetPoint_
				)
				expect(mediaQuery).toBe(answer[i])
			})
		})
	})
})
