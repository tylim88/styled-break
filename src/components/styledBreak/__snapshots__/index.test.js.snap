// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test core API test cssR test snapshot 1`] = `
@media (min-width:0px) {
  .c0 {
    color: red;
  }
}

@media (min-width:768px) and (max-width:1199.98px) {
  .c0 {
    color: yellow;
  }
}

<button
  className="c0"
/>
`;

exports[`test core API test styleHOC test snapshot 1`] = `
.c0 {
  color: blue;
  width: 150px;
}

@media (min-width:0px) {
  .c0 {
    color: red;
  }
}

@media (min-width:768px) and (max-width:1199.98px) {
  .c0 {
    color: yellow;
  }
}

@media (min-width:1200px) {
  .c0 {
    width: 250px;
  }
}

@media (min-width:768px) and (max-width:1199.98px) {
  .c0 {
    width: 350px;
  }
}

<button
  className="c0"
  color="yellow"
  width={50}
/>
`;

exports[`test core API test styledR test snapshot 1`] = `
.c0 {
  color: blue;
  width: 150px;
}

@media (min-width:0px) {
  .c0 {
    color: red;
  }
}

@media (min-width:768px) and (max-width:1199.98px) {
  .c0 {
    color: yellow;
  }
}

@media (min-width:1200px) {
  .c0 {
    width: 250px;
  }
}

@media (min-width:768px) and (max-width:1199.98px) {
  .c0 {
    width: 350px;
  }
}

<button
  className="c0"
  color="red"
  width={50}
/>
`;
